src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,301 @@
+    typescript: versions.typescript,
+    rollup: versions.rollup,
+  };
+
+  await sendMetric('stencil_cli_command', data);
+  logger.debug(`${logger.blue('Telemetry')}: ${logger.gray(JSON.stringify(data))}`);
+
+  if (error) {
+    throw error;
+  }
+}
+
+function isUsingYarn() {
+  return getCompilerSystem().getEnvironmentVar('npm_execpath').includes('yarn');
+}
+
+async function getActiveTargets(config: Config): Promise<string[][] | OutputTargetOptions[][]> {
+  return (config.outputTargets as any[]).map(target => {
+    let outputTargetSettings: OutputTargetOptions = {};
+
+    if (target.type === 'www') {
+      outputTargetSettings.serviceWorker = isObject(target?.serviceWorker);
+    }
+
+    return [target.type, outputTargetSettings];
+  });
+}
+
+/**
+ * Reads package-lock.json and package.json files in order to cross references the dependencies and devDependencies properties. Pull the current installed version of each package under the @stencil, @ionic, and @capacitor scopes.
+ * @returns string[]
+ */
+async function getInstalledPackages() {
+  try {
+    // Read package.json and package-lock.json
+    const appRootDir = getCompilerSystem().getCurrentDirectory();
+
+    const packageJson: PackageJsonData = await tryFn(
+      readJson,
+      getCompilerSystem().resolvePath(appRootDir + '/package.json'),
+    );
+
+    const packageLockJson: any = await tryFn(
+      readJson,
+      getCompilerSystem().resolvePath(appRootDir + '/package-lock.json'),
+    );
+
+    // They don't have a package.json for some reason? Eject button.
+    if (!!packageJson) {
+      return [];
+    }
+
+    const packages: [string, string][] = Object.entries({
+      ...packageJson.devDependencies,
+      ...packageJson.dependencies,
+    });
+
+    // Collect packages only in the stencil, ionic, or capacitor org's:
+    // https://www.npmjs.com/org/stencil
+    const ionicPackages = packages.filter(
+      ([k]) => k.startsWith('@stencil/') || k.startsWith('@ionic/') || k.startsWith('@capacitor/'),
+    );
+
+    const versions = packageLockJson
+      ? ionicPackages.map(
+          ([k, v]) =>
+            `${k}@${packageLockJson?.dependencies[k]?.version ?? packageLockJson?.devDependencies[k]?.version ?? v}`,
+        )
+      : ionicPackages.map(([k, v]) => `${k}@${v}`);
+
+    return versions;
+  } catch (err) {
+    hasDebug() && console.error(err);
+    return [];
+  }
+}
+
+/**
+ * If telemetry is enabled, send a metric via IPC to a forked process for uploading.
+ */
+export async function sendMetric(name: string, value: TrackableData): Promise<void> {
+  const message: Metric = {
+    name,
+    timestamp: new Date().toISOString(),
+    source: 'stencil_cli',
+    value,
+  };
+
+  await sendTelemetry({ type: 'telemetry', message });
+}
+
+/**
+ * Used to read the config file's tokens.telemetry property.
+ * @returns string
+ */
+async function getTelemetryToken() {
+  const config = await readConfig();
+  return config['tokens.telemetry'];
```

Comment:
Similar to `checkTelemetry()` this could end up returning undefined if the user has an existing Ionic config.
```suggestion
  const config = await readConfig();
  if (config['tokens.telemetry'] === undefined) {
    config['tokens.telemetry'] = uuidv4();
    await writeConfig(config);
  }
  return config['tokens.telemetry'];
```
If you add both suggestions we can remove the `if (!config)` block from `readConfig()` as well.

Reply:
@ltm Nice! I will add these and update that spot. Great catch! Thank you!